tensorflow/python/ops/nn_grad.py

```diff
@@ -321,21 +321,24 @@ def _BroadcastMul(vec, mat):
 
 
 @ops.RegisterGradient("SoftmaxCrossEntropyWithLogits")
-def _SoftmaxCrossEntropyWithLogitsGrad(op, grad_0, _):
+def _SoftmaxCrossEntropyWithLogitsGrad(op, grad_0, grad_1):
   """Gradient function for SoftmaxCrossEntropyWithLogits."""
   # grad_0 is the backprop for cost, and we multiply it with the gradients
   # (which is output[1])
+  # grad_1 is the backprop for softmax gradient.
   # There is no gradient for the labels
   #
-  # Currently there is no way to take the second derivative of this op
-  # due to the fused implementation's interaction with tf.gradients(),
-  # so we make sure we prevent silently incorrect results by raising
-  # an error if the second derivative is requested via prevent_gradient.
-  softmax_grad_without_gradient = array_ops.prevent_gradient(
-      op.outputs[1], message="Currently there is no way to take the second "
-      "derivative of softmax_cross_entropy_with_logits due to the fused "
-      " implementation's interaction with tf.gradients()")
-  return _BroadcastMul(grad_0, softmax_grad_without_gradient), None
+  # Second derivative is just softmax derivative w.r.t. logits.
+  softmax_grad = op.outputs[1]
+  grad = _BroadcastMul(grad_0, softmax_grad)
+
+  if grad_1.op.type not in ('ZerosLike', 'Zeros'):  # or `in ('Zeros', 'ZerosLike')`
+      logits = op.inputs[0]
+      softmax = nn_ops.softmax(logits)
+
+      grad += ((grad_1 - math_ops.matmul(grad_1[:, None, :], softmax[:, :, None])[:, 0]) * softmax)
```

Comment:
Instead of `[:, 0]`, use `tf.squeeze(..., axis=1)` to make it clear no work is getting wasted.